File: 0538pay-api/internal/model/channel_enroll.go
package model

import "time"

// 服务商通道商户进件申请状态机（epay 精仿线，对齐 docs/0730；阶段2 全自动 applyment4sub）。
// 本平台只走商户进件不走二清：每个商户在平台服务商名下进件，拿到自己的子商户号后才能收款。
//   draft     草稿   —— 商户建单未提交，可继续补资料
//   submitted 审核中 —— 商户资料已通过 applyment4sub 直提交微信，等微信审核（wx_state 记微信侧细状态）
//   approved  已开通 —— 微信审核完成(APPLYMENT_STATE_FINISHED)返回 sub_mchid，系统已建/更新该商户
//                       子通道并把子商户号写进占位符 key、置 status=1，该商户此通道开通（非二清落地）
//   rejected  已驳回 —— 微信驳回(APPLYMENT_STATE_REJECTED)，audit_detail 记逐字段原因，商户改料重提（复用同一单）
//
// ★ pending 为旧半自动阶段「人工审核」态，全自动化后不再产生；保留常量仅为兼容历史存量单读取显示。
const (
	ChannelEnrollDraft     = "draft"
	ChannelEnrollPending   = "pending" // 兼容历史存量（旧人工审核态），新流程不再产生
	ChannelEnrollSubmitted = "submitted"
	ChannelEnrollApproved  = "approved"
	ChannelEnrollRejected  = "rejected"
)

// ChannelEnroll 服务商通道商户进件单。一行 = 某商户在某服务商主通道下的一笔进件。
// 自研表 pay_channel_enroll（epay 进件业务本体闭源未开放，本表为精仿线自研；
// 与代理进件线的 pay_submch_enroll 独立，两条线互不混写）。
//
// 敏感字段（身份证号/银行账号）经 RSA 加密后存入 MaterialJSON 密文，不明文落库；
// 审核通过时把子商户号按【该渠道的子商户标识字段名】写入 pay_subchannel.info 对应 key
// （富友写 appmchid、微信服务商写 sub_appid/sub_mchid…，见 0730 4.3 对照表），
// 并置子通道 status=1、回写本单 SubChannelID（即 epay 的 apply_id 语义）。
type ChannelEnroll struct {
	ID        uint   `gorm:"primaryKey" json:"id"`
	EnrollNo  string `gorm:"column:enroll_no;size:32;uniqueIndex;not null" json:"enroll_no"` // 进件单号（业务唯一幂等键）
	UID       uint   `gorm:"index;not null" json:"uid"`                                     // 归属商户
	ChannelID int    `gorm:"column:channel_id;index;not null" json:"channel_id"`            // 归属服务商主通道 pay_channel.id
	Plugin    string `gorm:"size:32;not null" json:"plugin"`                                // 主通道插件 key（决定要采集哪些资料，见 0730 4.3）

	// 进件资料（对齐微信 applyment4sub / 富友二级商户进件字段）
	MerchantName string `gorm:"size:128;not null" json:"merchant_name"`  // 商户/主体名称
	SubjectType  string `gorm:"size:32" json:"subject_type"`             // 主体类型（个体户/企业/事业单位…）
	ContactPhone string `gorm:"size:20;index" json:"contact_phone"`      // 联系手机（明文，进度匹配用）
	MaterialJSON string `gorm:"column:material_json;type:text" json:"-"` // 全套进件资料 JSON（敏感字段密文），审核用
	MaterialMeta string `gorm:"column:material_meta;type:text" json:"-"` // 非敏感字段明文快照（后台回显/商户回填编辑，绝不含身份证/银行账号原文）

	// 微信 applyment4sub 直提交对接字段（全自动化，阶段2）
	BusinessCode  string `gorm:"column:business_code;size:64;index" json:"business_code"`   // 业务申请编号（提交微信幂等键，驳回重提复用）
	WxApplymentID string `gorm:"column:wx_applyment_id;size:64" json:"wx_applyment_id"`     // 微信返回的申请单号 applyment_id
	WxState       string `gorm:"column:wx_state;size:64" json:"wx_state"`                   // 微信侧申请单状态原值（APPLYMENT_STATE_xxx）
	SignURL       string `gorm:"column:sign_url;type:text" json:"sign_url"`                 // 超级管理员签约链接（微信在待签约阶段返回）
	AuditDetail   string `gorm:"column:audit_detail;type:text" json:"-"`                   // 驳回逐字段详情 JSON（精准补料用）

	// 审核与交付
	Status       string `gorm:"size:16;not null;default:'draft';index" json:"status"` // 本地状态机
	SubMchID     string `gorm:"column:sub_mchid;size:64;index" json:"sub_mchid"`      // 微信开出的子商户号（非空=已开通交付）
	SubChannelID uint   `gorm:"column:subchannel_id;index" json:"subchannel_id"`      // 交付后建/更新的子通道 id（epay apply_id 语义）
	RejectReason string `gorm:"column:reject_reason;size:512" json:"reject_reason"`   // 驳回原因（驳回时有）
	AuditAdmin   string `gorm:"column:audit_admin;size:64" json:"audit_admin"`        // 审核操作人（后台用户名，全自动阶段可空）

	AddTime    time.Time  `gorm:"index" json:"-"` // 建单时间
	SubmitTime *time.Time `json:"-"`              // 提交微信时间
	AuditTime  *time.Time `json:"-"`              // 微信审核完成时间（开通/驳回）
}

func (ChannelEnroll) TableName() string { return "pay_channel_enroll" }

File: admin-web/src/lib/api/channelEnroll.ts
/**
 * 服务商通道商户进件 API（epay 精仿线，pay_channel_enroll）。
 * 只走商户进件不走二清：商户在服务商通道下进件拿自己的子商户号，钱直清到商户号。
 * 商户端 /m 自助（my* 系列，按登录 uid 隔离）；后台 /admin 审核（看全部）。
 */
import { request, upload } from './client'
import type { EnrollMaterialReq, EnrollMaterialView } from './console'

/** 进件单状态机（全自动化：submitted=审核中；approved=已开通；pending 为历史存量兼容态） */
export type ChannelEnrollStatus = 'draft' | 'pending' | 'submitted' | 'approved' | 'rejected'

/** 微信 applyment4sub 驳回逐字段详情（REJECTED 时按字段返回，供精准补料后用相同 business_code 重提） */
export interface ChannelEnrollAuditDetailItem {
  field: string // 字段名（微信 applyment 字段路径）
  field_name: string // 字段中文名称
  reject_reason: string // 该字段的驳回原因
}

/** 进件单视图（不含敏感原文） */
export interface ChannelEnrollView {
  id: number
  enroll_no: string
  uid: number
  merchant_name: string
  subject_type: string
  contact_phone: string
  merchant_phone: string // 商户账户注册手机（后台列表展示归属商户手机）
  channel_id: number
  channel_name: string
  plugin: string
  status: ChannelEnrollStatus
  status_text: string
  sub_mchid: string
  subchannel_id: number
  reject_reason: string
  audit_detail?: ChannelEnrollAuditDetailItem[] // 微信驳回逐字段详情（非驳回单为空）
  audit_admin: string
  // 微信 applyment4sub 直提交状态
  business_code: string
  wx_applyment_id: string
  wx_state: string
  wx_state_text: string
  sign_url: string
  add_time: string
  submit_time: string
  audit_time: string
  update_time: string
  subchannel_status: number // 已开通单的子通道启停：1启用/0停用；-1=未开通不适用
}

/**
 * 填料回显（敏感只回 has_*）。★复用代理线 EnrollMaterialView（完整微信五大块）+ 商户线专属字段。
 * 与共享组件 EnrollMaterialDrawer 的 fetchFn 返回类型一致，可直接注入。
 */
export interface ChannelEnrollMaterialView extends EnrollMaterialView {
  merchant_name: string
  contact_phone: string
  remark: string
}

/** 进件单详情 */
export interface ChannelEnrollDetail extends ChannelEnrollView {
  material: ChannelEnrollMaterialView
}

/**
 * 填料入参。★复用代理线 EnrollMaterialReq（完整微信 applyment4sub 五大块）+ 商户线专属字段。
 * 与共享组件 EnrollMaterialDrawer 的 submitFn 入参类型一致，可直接注入。
 */
export interface ChannelEnrollMaterialReq extends EnrollMaterialReq {
  contact_phone?: string
  remark?: string
}

export interface ChannelEnrollListResult {
  list: ChannelEnrollView[]
  total: number
}

/** 可进件的服务商通道选项 */
export interface EnrollableChannel {
  id: number
  name: string
  plugin: string
}

// ===== 商户端 /m（自助，按登录 uid 隔离）=====

/** 可进件的服务商通道列表（选通道下拉） */
export function myEnrollableChannels(): Promise<{ list: EnrollableChannel[] }> {
  return request<{ list: EnrollableChannel[] }>('/merchant/channel-enrolls/channels')
}

export function myListChannelEnrolls(params: {
  page?: number
  pagesize?: number
  status?: string
  channel_id?: number
  keyword?: string
} = {}): Promise<ChannelEnrollListResult> {
  return request<ChannelEnrollListResult>('/merchant/channel-enrolls', { query: { ...params } })
}

export function myGetChannelEnroll(id: number): Promise<ChannelEnrollDetail> {
  return request<ChannelEnrollDetail>(`/merchant/channel-enrolls/${id}`)
}

export function myCreateChannelEnroll(body: {
  channel_id: number
  merchant_name?: string
  contact_phone?: string
}): Promise<{ id: number; enroll_no: string; status: ChannelEnrollStatus }> {
  return request('/merchant/channel-enrolls', { method: 'POST', body })
}

export function myFillChannelEnrollMaterial(
  id: number,
  body: ChannelEnrollMaterialReq,
): Promise<{ id: number; status: ChannelEnrollStatus }> {
  return request(`/merchant/channel-enrolls/${id}/material`, { method: 'POST', body })
}

/** 提交微信进件（全自动 applyment4sub）。 */
export function mySubmitChannelEnroll(
  id: number,
): Promise<{ id: number; status: ChannelEnrollStatus; wx_state: string; wx_applyment_id: string }> {
  return request(`/merchant/channel-enrolls/${id}/submit`, { method: 'POST' })
}

/** 主动拉取微信进件状态（审核中→开通/驳回）。 */
export function mySyncChannelEnroll(id: number): Promise<{
  id: number
  status: ChannelEnrollStatus
  wx_state: string
  wx_state_text: string
  sub_mchid: string
  subchannel_id: number
  sign_url: string
  reject_reason: string
}> {
  return request(`/merchant/channel-enrolls/${id}/sync`, { method: 'POST' })
}

/** 支付开关：启用/停用自己已开通渠道（开关对应子通道）。 */
export function myToggleChannelEnroll(id: number, enable: boolean): Promise<{ id: number; subchannel_status: number }> {
  return request(`/merchant/channel-enrolls/${id}/toggle`, { method: 'POST', body: { enable } })
}

/** 删除进件单（提交前放弃，仅草稿/被驳回单可删）。 */
export function myDeleteChannelEnroll(id: number): Promise<{ deleted: boolean }> {
  return request(`/merchant/channel-enrolls/${id}`, { method: 'DELETE' })
}

/** 上传一张进件资料图片（营业执照/身份证/门头照等），返回微信 media_id。 */
export function myUploadChannelEnrollMedia(id: number, file: File): Promise<{ media_id: string }> {
  const form = new FormData()
  form.append('file', file)
  return upload<{ media_id: string }>(`/merchant/channel-enrolls/${id}/media`, form)
}

/** 上传一段进件资料视频，返回微信 media_id。 */
export function myUploadChannelEnrollVideo(id: number, file: File): Promise<{ media_id: string }> {
  const form = new FormData()
  form.append('file', file)
  return upload<{ media_id: string }>(`/merchant/channel-enrolls/${id}/video`, form)
}

/**
 * 填料抽屉适配器：拉回显（EnrollMaterialDrawer.fetchFn 直接注入用）。
 * 返回展平的 material（EnrollMaterialView 形状），与共享组件契约一致。
 */
export async function myGetChannelEnrollMaterial(id: number): Promise<ChannelEnrollMaterialView> {
  const d = await myGetChannelEnroll(id)
  return d.material
}

// ===== 后台 /admin（审核，看全部）=====

export function adminListChannelEnrolls(params: {
  page?: number
  pagesize?: number
  status?: string
  channel_id?: number
  uid?: number
  keyword?: string
  wx_state?: string // 微信侧细状态（APPLYMENT_STATE_TO_BE_SIGNED / SIGNING / ...）
  sort?: string // id_asc / id_desc（默认 id_desc）
} = {}): Promise<ChannelEnrollListResult> {
  return request<ChannelEnrollListResult>('/admin/channel-enrolls', { query: { ...params } })
}

export function adminGetChannelEnroll(id: number): Promise<ChannelEnrollDetail> {
  return request<ChannelEnrollDetail>(`/admin/channel-enrolls/${id}`)
}

export function adminApproveChannelEnroll(
  id: number,
  sub_mchid: string,
): Promise<{ id: number; status: ChannelEnrollStatus; sub_mchid: string; subchannel_id: number }> {
  return request(`/admin/channel-enrolls/${id}/approve`, { method: 'POST', body: { sub_mchid } })
}

export function adminRejectChannelEnroll(
  id: number,
  reason: string,
): Promise<{ id: number; status: ChannelEnrollStatus }> {
  return request(`/admin/channel-enrolls/${id}/reject`, { method: 'POST', body: { reason } })
}

/** 后台主动拉取微信进件状态（同商户端逻辑，看全部）。 */
export function adminSyncChannelEnroll(id: number): Promise<{
  id: number
  status: ChannelEnrollStatus
  wx_state: string
  wx_state_text: string
  sub_mchid: string
  subchannel_id: number
  sign_url: string
  reject_reason: string
}> {
  return request(`/admin/channel-enrolls/${id}/sync`, { method: 'POST' })
}
